cpp/src/utilities/type_dispatcher.hpp

```diff
@@ -16,20 +16,186 @@
 #ifndef TYPE_DISPATCHER_HPP
 #define TYPE_DISPATCHER_HPP
 
-#include "wrapper_types.hpp"
-#include "release_assert.cuh"
+#include <utilities/wrapper_types.hpp>
+#include <utilities/release_assert.cuh>
 
 #include <cudf/types.h>
 
 // Forward decl
 class NVStrings;
 
-#include <cassert>
 #include <utility>
+#include <type_traits>
+
+namespace cudf {
+
+namespace detail {
+
+template <typename... List> struct list { };
+
+template < typename Tp, typename... List >
+struct first_non_matching;
+
+template <typename T> struct non_match { using type = T; };
+
+template < typename Tp, typename Head, typename... Rest >
+struct first_non_matching<Tp, Head, Rest...> :
+    std::conditional<not std::is_same<Tp, Head>::value, non_match<Head>, first_non_matching<Tp, Rest...> >::type
+{};
+
+template < typename Tp >
+struct first_non_matching<Tp> { }; // no type set
+
+using all_column_element_types =
+    list<
+        int8_t, int16_t, int32_t, int64_t, float, double, cudf::date32, cudf::date64,
+        cudf::timestamp, cudf::category, cudf::nvstring_category
+    >;
+
+
+template <typename>
+struct sfinae_true : std::true_type{};
+
+// Note: The .template is known as a "template disambiguator"
+// See here for more information: https://stackoverflow.com/questions/3786360/confusing-template-error
+template <class F, typename T, typename... Us>
+static auto test_templated_invoke_operator(int) ->
+    sfinae_true<decltype(std::declval<F>().template operator()<T>(std::declval<Us>()... ))>;
+
+template <class, typename, typename... Us>
+static auto test_templated_invoke_operator(long) -> std::false_type;
+
+// A useful trait in itself:
+template <class F, typename T, typename... Us>
+struct has_templated_invoke_operator : decltype( test_templated_invoke_operator<F, T, Us...>(int{}) )
+{ };
+
+template <bool ActuallyInvoke, typename R, class F, typename T, typename... Ts>
+struct invoke_if_possible_inner;
+
+template <class F, typename R, typename T, typename... Ts>
+struct invoke_if_possible_inner<false, R, F, T, Ts...>
+{
+    constexpr __host__ __device__
+    R operator()(F functor, Ts&&... params) {
+        // throw std::invalid_argument("Unsupported column element type");
+        return R();
+    }
+};
+
+template <class F, typename R, typename T, typename... Ts>
+struct invoke_if_possible_inner<true, R, F, T, Ts...>
+{
+// Note that some functors may only have their operator()'s defined for host-side code,
+// and other only for device-side code; but - since these decorators are not really part of
+// a function's C++ type, we can't quite work our way around this issue, and have to
+// result to the following kludge:
+#pragma hd_warning_disable
```

Comment:
Please also include `#pragma nv_exec_check_disable`, this is necessary.

Reply:
@jrhemstad : Are you sure it's necessary? What is the difference between these two pragmas?

... perhaps this is a good time to mention that AFAICT the pragmas are not listed and documented anywhere that I've seen.

Reply:
`nv_exec_check_disable` is newer and unsupported by older `nvcc` (unsure which version). As such, I suspect we can probably replace `hd_warning_disable` entirely.

I have also experienced that there were compiler warnings that still made it through with using `hd_warning_disable`.

Reply:
Ok, I'll just add it then.

Reply:
This isn't addressed yet, so I unresolved.

Reply:
Oh, sorry, I forgot to commit the extra pragma.